File: EASY/Easy_Cpp/intersection_of_two_linked_lists.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

# Solution1: using hash set
class Solution1:
    def getIntersectionNode(self, headA: ListNode, headB: ListNode) -> Optional[ListNode]:
        visited = set()
        while headA:
            visited.add(headA)
            headA = headA.next
        while headB:
            if headB in visited:
                return headB
            headB = headB.next
        return None

# Soltion2: without using extra memory

class Solution2:
    def getIntersectionNode(self, headA: ListNode, headB: ListNode) -> Optional[ListNode]:
        l1, l2 = headA, headB
        while l1 != l2: # if not pointing to the same node move both l1 and l2 move to the next node of their linked lists
            l1 = l1.next if l1 else headB 
            l2 = l2.next if l2 else headA

        return l1

File: EASY/Easy_Cpp/intersection_of_two_linked_lists.cpp
class Solution {
public:
    ListNode *get intersectionNode(ListNode *headA, ListNode *headB){
        ListNode *l1 = headA;
        ListNode *l2 = headB;
        while (l1 != l2){
            l1 = l1 ? l1->next : headB;
            l2 = l2 ? l2->next : headA;
        }
        return l1;
    }
};